### ISSUE TITLE
Ergonomic rename of zip entries

### ISSUE THREAD

**Issue by zoj613** [dbuenzli/zipc#8]
I have a workflow where I attempt to rename zip file entries opened using this library and then write them to disk like so:
```ocaml
  	module M = Map.Make(String)

	type t = Zipc.t

    let rename t ok nk =
      let m = Zipc.to_string_map t in
      let m1, m2 = M.partition (fun k _ -> String.starts_with ~prefix:ok k) m in
      let l = String.length ok in
      let s = Seq.map
        (fun (k, v) -> nk ^ String.(length k - l |> sub k l), v) @@ M.to_seq m1 in
      Zipc.of_string_map @@ M.add_seq s m2
```
I was surprised to learn that the update zip file did not actually contain the entries with updated paths. In order for it to work, it seems like I had to update the above example by manually creating new zip member types to include in the resulting `Seq` type. I had the impression that when constructing a new `Zipc.t` type using a `Seq` with updated path values the new path would be included as part of the new Zipc.t.

```ocaml
    let rename t ok nk =
      let m = Zipc.to_string_map t in
      let m1, m2 = M.partition (fun k _ -> String.starts_with ~prefix:ok k) m in
      let l = String.length ok in
      let s = Seq.map
        (fun (k, v) ->
          let mtime = Zipc.Member.mtime v in
          let mode = Zipc.Member.mode v in
          let kind = Zipc.Member.kind v in
          let path = nk ^ String.(length k - l |> sub k l) in
          match Zipc.Member.make ~mtime ~mode ~path kind with
          | Error e -> failwith e
          | Ok member -> path, member) @@ M.to_seq m1 in
      Zipc.of_string_map @@ M.add_seq s m2
```

My question is: Is there a simpler way of achieving the above without having to copy over all attributes of the current `Zipc.Member.t` and creating new ones with an updated `path` value? The current solution seems like a lot of work for just updating a path. It would be nice to have a function like `update_path`. What are your thoughts?

**Comment by dbuenzli** [comment]
> I was surprised to learn that the update zip file did not actually contain the entries with updated paths.

Ah I knew I should not have provided these convenience functions to a non-normalized abstraction. It's written [here](https://github.com/dbuenzli/zipc/blob/48464a2bfe608e65282a2534f4fc3751e5fa8ed6/src/zipc.mli#L321-L325) that what you do won't work. But no one reads the docs.

The reason I did provide this that I didn't want to replicate all `Map` functions in `Zipc` but still allow people to, for example, easily merge two archives. I should have called the function `Zipc.unsafe_of_string_map`.

> The current solution seems like a lot of work for just updating a path. 

Well one one hand it's just three obvious `let`s. 

On the other hand what worries me most is the *tons* of useless computation work you are doing. I'm not sure exactly how came up with the idea of converting to maps, partition them and then iterate over them with `Seq`s (`Seq`s are always code smells). At that point you have gone twice over the member of your archive, have sorted its paths twice, redone the map twice, and generated a lot of gc garbage with totally useless Seqs. Intead of a simple `Zipc.fold`.

While I'm here also please don't make gratuitous use of `|>` they make the code harder to read: if I see a `String.sub` I mostly remember how to read it despite the lack of labels but not *obviously* if its last argument comes from the left! Don't make me think!

So I hope you won't take the above badly but the function you wanted to write is:

```ocaml
let rename z ~prefix ~new_prefix =
  let rename_member ~prefix ~new_prefix m =
    let path = Zipc.Member.path m in
    if not (String.starts_with ~prefix path) then m else
    let l = String.length prefix in
    let path = new_prefix ^ String.sub path l (String.length path - l) in
    let mtime = Zipc.Member.mtime m in
    let mode = Zipc.Member.mode m in
    let kind = Zipc.Member.kind m in
    match Zipc.Member.make ~mtime ~mode ~path kind with
    | Ok m -> m | Error e -> failwith e
  in
  let add m acc = Zipc.add (rename_member ~prefix ~new_prefix m) acc in
  try Ok (Zipc.fold add z Zipc.empty) with Failure e -> Error e
```

Now yes it would likely be nice to have an `update_path` function but then it would also be nice to have for example `update_{mode,mtime}`. We could have an 

```ocaml
Member.with' : ?mtime:… -> ?path:… -> … -> Member.t -> Member.t 
```
but on the other hand it seems [I wasn't more disturbed](https://github.com/dbuenzli/zipc/blob/48464a2bfe608e65282a2534f4fc3751e5fa8ed6/test/zipc_tool.ml#L456-L458) by that by the above code. I'm pretty sure I contemplated adding it before thinking it was not worth the hassle.

But we can keep it open to see if more people want this.

**Comment by zoj613** [comment]
You are absolutely right that I should have used fold; and I did in the first iteration of the code, but because I could not successfully rename the entries I resorted to the current approach using `Seq` thinking it was the correct way. Thanks, I will update the code based on your suggestions.

Regarding the use of `|>`, It was an attempt to be concise so I don't end up repeating `String.`. But I agree that this approach sacrificed readability for conciseness, which I shouldn't have prioritized.

In your proposed version above, you use 2 closures (I assume to aid readability) instead of "inlining" the code inside the
```ocaml
Zipc.fold
  (fun m acc ->
     ...) m Zipc.empty
```
pattern. Is there a reason to prefer the former to the latter, considering that the former leads to more allocations? (Please correct me if i'm wrong in assuming closures require more allocations that lambdas.

**Comment by dbuenzli** [comment]
> (Please correct me if i'm wrong in assuming closures require more allocations that lambdas.

A closure is a function with free variables (variables not bound by the function arguments). If your anonymous functions ("lambdas") have free variables they are closures. For example every anonymous function of your initial message are also closures. So your assumption is a bit non sensical.

In general it's better to name your functions and expressions and avoid the "inline" style as soon as it doesn't fit on the line. The "inline" style (which is sadly very widespread in OCaml code) requires too much labor from readers to deconstruct and delimit expressions. You want to facilitate their task by meaningfully naming them in bite-sized expressions.

Regarding the fact that there are two closures, I wouldn't worry to much about it; and the compiler may inline stuff. Here it was done because `rename_member` is a good candidate to become a toplevel function to be reused on individual member. In fact initially I had only `add`:

```ocaml
let add m acc = 
  let m =
    let path = Zipc.Member.path m in
    if not (String.starts_with ~prefix path) then m else
    let l = String.length prefix in
    …
  in 
  Zipc.add m acc 
in
```
but I thought the meaning was expressed more meaningfully by separating the `add` and the `rename` function.

**Comment by zoj613** [comment]
> > (Please correct me if i'm wrong in assuming closures require more allocations that lambdas.
> 
> A closure is a function with free variables (variables not bound by the function arguments). If your anonymous functions ("lambdas") have free variables they are closures. For example every anonymous function of your initial message are also closures. So your assumption is a bit non sensical.
> 
> In general it's better to name your functions and expressions and avoid the "inline" style as soon as it doesn't fit on the line. The "inline" style (which is sadly very widespread in OCaml code) requires too much labor from readers to deconstruct and delimit expressions. You want to facilitate their task by meaningfully naming them in bite-sized expressions.
> 
> Regarding the fact that there are two closures, I wouldn't worry to much about it; and the compiler may inline stuff. Here it was done because `rename_member` is a good candidate to become a toplevel function to be reused on individual member. In fact initially I had only `add`:
> 
> ```ocaml
> let add m acc = 
>   let m =
>     let path = Zipc.Member.path m in
>     if not (String.starts_with ~prefix path) then m else
>     let l = String.length prefix in
>     …
>   in 
>   Zipc.add m acc 
> in
> ```
> 
> but I thought the meaning was expressed more meaningfully by separating the `add` and the `rename` function.

Ok got it. Thanks a lot, this is valuable advice for me and will help me adopt good habits for any future development work since I have only been using OCaml for about 6 months now.

**Comment by dbuenzli** [comment]
You are welcome. 

Oh and by the way if you go back to my code, in fact there is only one closure, not two as you suggested. `rename_member` has no free variables it's a function, so a closure is only created for the `add` function (which has the `~prefix` and `~new_prefix` free variables).
